src/xmipp/libraries/reconstruction_starpu/reconstruct_fourier_codelet_load_projections.cpp

```diff
@@ -0,0 +1,108 @@
+/***************************************************************************
+ *
+ * Authors:     Jan Polak (456647@mail.muni.cz)
+ *              (some code derived from other Xmipp programs by other authors)
+ *
+ * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ * 02111-1307  USA
+ *
+ *  All comments concerning this program package may be sent to the
+ *  e-mail address 'xmipp@cnb.csic.es'
+ ***************************************************************************/
+
+#include <iostream>
+
+#include <core/args.h>
+#include <core/metadata.h>
+#include <data/projection.h>
+#include <core/xmipp_fftw.h>
+
+#include "reconstruct_fourier_codelets.h"
+#include "reconstruct_fourier_starpu_util.h"
+
+void func_load_projections(void* buffers[], void* cl_arg) {
+	const LoadProjectionArgs& arg = *static_cast<LoadProjectionArgs*>(cl_arg);
+
+	uint32_t& amountLoaded = ((LoadedImagesBuffer*) STARPU_VARIABLE_GET_PTR(buffers[0]))->noOfImages;
+	float* outImageData = (float*)STARPU_VECTOR_GET_PTR(buffers[1]);
+	const size_t outImageDataStride = STARPU_VECTOR_GET_ELEMSIZE(buffers[1]) / sizeof(float);
+	RecFourierProjectionTraverseSpace* outSpaces = (RecFourierProjectionTraverseSpace*)STARPU_VECTOR_GET_PTR(buffers[2]);
+
+	ApplyGeoParams geoParams;
+	geoParams.only_apply_shifts = true;
+
+	MultidimArray<float> paddedImageData; // Declared here so that internal allocated memory can be reused
+
+	uint32_t traverseSpaceIndex = 0;
+	uint32_t projectionIndex = 0;
+
+	for (uint32_t projectionInBatch = arg.batchStart; projectionInBatch < arg.batchEnd; projectionInBatch++) {
+		const size_t imageObjectIndex = arg.imageObjectIndices[projectionInBatch];
+
+		//Read projection from selfile, read also angles and shifts if present
+		//but only apply shifts (set above)
+		// FIXME following line is a current bottleneck, as it calls BSpline interpolation
+		Projection proj;
+		proj.readApplyGeo(arg.selFile, imageObjectIndex, geoParams);
```

Comment:
The bottleneck comment is old, from the cuda version, it may even be yours. I haven't looked much into what this actually does. But if the comment is true and most of the time in this function is indeed spent in this call, then yes, GPU implementation would help a lot. (It was actually on my TODO list to investigate this, so thanks for the comment.)

Reply:
Ok, if you find out that this is a problem, let me know.